### PR TITLE
feat: Add `QueuedRequestController` batching

### PR DESCRIPTION
**Pull request by Gudahtt**
## **Description**

The `QueuedRequestController` will now batch requests by origin. Requests from the same UI now get shown together, just as they were before this queuing feature was added. The only remaining difference is that we no longer show requests from different origins in the same batch, which we view as a security/usability improvement.

## **Related issues**

Fixes:

## **Manual testing steps**

**important** make sure any dapp used for testing has a permission (eth_accounts for example), as well as the per-domain network feature flag turned on, then refresh the dapp before begining testing.

make sure to also test that everything works unchanged with the feature flag off

1. Turn on per-domain network feature flag
2. go to test dapp and connect
3. refresh the test dapp
4. go to any other website (msn.com for example)
5. in the devtools, run this:
```
await window.ethereum.request({
  "method": "wallet_requestPermissions",
  "params": [
    {
      "eth_accounts": {}
    }
  ]
});
```
6. go back to the testdapp tab, make 2 or 3 send transaction calls
7. ensure that the UI displays the number of transactions being processed, and allows you to go back and forward through the transactions.
8. before confirming or rejecting any of those transactions, switch to the tab with msn.ca and run this 2-3 times:
```
await window.ethereum.request({
  "method": "eth_sendTransaction",
  "params": [
    {
      "to": "0x77ff877eE79c3f8C5aD244c037e5580EdCC6eF2c",
      "from": "your address",
      "value": "0x1"
    }
  ]
});
```
9. ensure that the you are still only able to scrub through the 3 transactions. Reject each of them. 
10. You should now see the next set of 3 (from the domain msn.com).

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by Gudahtt** [comment]
@metamaskbot update-policies

**Comment by Gudahtt** [comment]
@metamaskbot update-policies

**Comment by Gudahtt** [comment]
@metamaskbot update-policies

**Comment by Gudahtt** [comment]
@metamaskbot update-policies

**Comment by Gudahtt** [comment]
@metamaskbot update-policies

File: package.json
Comment (Gudahtt) [inline package.json:289]: This is resulting in a bunch of peer dependency warnings

TODO: Validate that they're safe to ignore, or bump the other packages as well in the same PR

File: test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js
Comment (Gudahtt) [inline test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js:1]: The test description here talks about showing network confirmations, but we don't show confirmations on switch anymore.

Perhaps we should update this test to show that the network is switching, and leave about the "confirmations" part.

Comment (Gudahtt) [reply test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js:1]: The same goes for the other request queuing test suite as well

**Comment by Gudahtt** [comment]
@metamaskbot update-policies

**Comment by Gudahtt** [comment]
@metamaskbot update-policies

**Comment by Gudahtt** [comment]
@metamaskbot update-policies

**Comment by BelfordZ** [comment]
Just noticed a bug:

1. turn on queuing feature flag
2. go to docs site https://docs.metamask.io/wallet/reference/ & clear permissions (disconnect from dapp)
3. check the docs selected network (call eth_chain id)
4. switch the wallet to a different network than the dapp
5. make a `wallet_switchEthereumChain` call from docs, switch to a chain that's not the same as the wallet's selected network, minimize the notification
6. make a eth_requestPermissions call for eth_accounts

-> double confirmation for switch, and the wallet ends up being switched to the docs' original selected network.

**Comment by BelfordZ** [comment]
problem 1: wallet_requestPermissions should not switch chain before executing

---

### problem 2: the networkClientId for a request that is queued *after* a switchEthereumChain request is incorrect

Consider the scenario where a dapp makes the following calls, in order, but without waiting for the response of one before sending the next (assume the dapp selected network is 0x5, and the wallet is on 0xa):
  1. `wallet_switchEthereumChain(0x1)`
  2. `wallet_requestPermissions({ eth_accounts: {} })`
In this scenario, what we expect to occur is:
1. confirmation dialog for switching from `0xa` to `0x1`
2. confirmation dialog for requesting the accounts
3. if the user opens the wallet, they will see that the network is '0x1'
4. if the dapp calls eth_chainId, they will get back '0x1'

What occurs atm though:
1. confirmation dialog for switching from `0xa` to `0x1`
2. confirmation dialog for switching from `0x1` to `0x5`
3. confirmation dialog for requesting the accounts
4. if the user opens the wallet, they will see that the network is '0x5'
5. if the dapp calls eth_chainId, they will get back '0x1'

**Comment by Gudahtt** [comment]
Rebased to resolve conflicts

**Comment by Gudahtt** [comment]
I believe I've found a bug related to our provider RPC pipeline and the SelectedNetworkController. It may be related to the bugs you've found here.

When testing this PR, I'd recommend doing a _hard-refresh_ of all dapps after granting or clearing permissions. That should work around the bug that I've found (will share more details shortly)

**Comment by jiexi** [comment]
Addresses problem 1:

https://github.com/MetaMask/core/pull/4066

**Comment by BelfordZ** [comment]
update: Ive re-tested the 2nd issue and it doesn't seem to be a problem anymore. Not sure what was going on before, but it seems to work as expected now. (just make sure the dapp has permissions, otherwise everything is pretty messed up)

**Comment by BelfordZ** [comment]
Rebased to resolve conflicts

**Comment by BelfordZ** [comment]
@metamaskbot update-policies

**Review by jiexi — DISMISSED** [review]

**Comment by BelfordZ** [comment]
Rebased again!

File: package.json
Comment (BelfordZ) [reply package.json:306]: Seems like theres only the one warning about queued request controller using a different version of selected network controller:

```
@metamask/selected-network-controller is listed by your project with version 9.0.0, which doesn't satisfy what @metamask/queued-request-controller (p81a9c) requests (^10.0.0).
```

the main change in 10.x.x is the update to base controller. The API afaik is the same, the typings have changed a bit here and there, but we arent getting any type errors so I think we are good to go.

https://github.com/MetaMask/core/blob/main/packages/selected-network-controller/CHANGELOG.md#1000

**Review by jiexi — APPROVED** [review]

**Review by BelfordZ — APPROVED** [review]

**Comment by Gudahtt** [comment]
I have re-reviewed the recent changes, LGTM!

**Pull request merged**
